poly-commitment/src/hash_map_cache.rs

```diff
@@ -0,0 +1,41 @@
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct HashMapCache<Key: Hash, Value> {
+    contents: Arc<Mutex<HashMap<Key, Value>>>,
+}
+
+impl<Key: Hash + std::cmp::Eq, Value> HashMapCache<Key, Value> {
+    pub fn new() -> Self {
+        HashMapCache {
+            contents: Arc::new(Mutex::new(HashMap::new())),
```

Comment:
Cool! I did something like this in parallel in my working branch: https://github.com/o1-labs/proof-systems/commit/444e07f4b19207096852e6d9beed09586dc471fd#diff-c89c45dc84ecd1f409b94d69ca374ed88678982f04f419b79435fd5ebff85b4fR362

Reply:
@mrmr1993 I'd suggest using `RWLock` though. Unless I'm misunderstanding, `Mutex` won't allow two concurrent reads, locking several parallel uses (which is the case in mina), and the upgrade to `RWLock` is easy. Any reasons not to?